src/librustc_typeck/collect.rs

```diff
@@ -2430,7 +2430,7 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
         } else if attr.check_name(sym::thread_local) {
             codegen_fn_attrs.flags |= CodegenFnAttrFlags::THREAD_LOCAL;
         } else if attr.check_name(sym::track_caller) {
-            if tcx.is_closure(id) || tcx.fn_sig(id).abi() != abi::Abi::Rust {
+            if !tcx.is_closure(id) && tcx.fn_sig(id).abi() != abi::Abi::Rust {
```

Comment:
Could these have comments stating that we error when using target feature on a closure in `check_attr`?

Reply:
Could this be `tcx.fn_sig(id).abi() != abi::Abi::Rust && tcx.fn_sig(id).abi() != abi::Abi::RustCall` instead? That would allow `#[track_caller]` on explicit `extern "rust-call" fn`.

Reply:
Does `#[track_caller]` work with the `rust-call` ABI?  I feel like that change isn't in the scope of this PR